V3

mahavrila opened the pull request:
V3 of SCAutolib differs from V2 mostly in:

- card objects are separated from user object
- many user attributes and methods are moved to card
- CAs are treated differently
- structures for physical cards are introduced, however it's still WIP

Users and cards are related only in prepare phase where objects are created. Card knows it's user. It's needed as some methods needs to know is user is local or kerberos user. Data of user and cards objects are stored separately.

Jakuje commented on SCAutolib/models/card.py:
```diff
@@ -275,11 +296,139 @@ def create(self):
+        :param exp_traceback: Traceback of the exception
+        """
+        if exp_type is not None:
+            logger.error("Exception in virtual smart card context")
```
```suggestion
            logger.error("Exception in physical smart card context")
```

x00Pavel commented on SCAutolib/controller.py:
```diff
@@ -171,11 +199,16 @@ def setup_local_ca(self, force: bool = False):
         dump_to_json(self.local_ca)
 
+    def setup_custom_ca(self, card_data: dict):
+        if card_data["card_type"] == "physical":
```
not sure what the flow for this method, but seems like card_data should be some object with `card_type` attribute

mahavrila replied:
Earlier in controller a .json file is deserialized  and provided as dictionary to _validate_configuraton that returns another dictionary  called 'self.lib_conf'. Part of this dictionary is is provided as a parameter here and I think approaching it as dict is correct here.

x00Pavel commented on SCAutolib/controller.py:
```diff
@@ -171,11 +199,16 @@ def setup_local_ca(self, force: bool = False):
+    def setup_custom_ca(self, card_data: dict):
+        if card_data["card_type"] == "physical":
+            ca = local_ca_factory(create=True, card_data=card_data)
+            dump_to_json(ca)
```
it would be nice to have some log messaga if CA wasn't created

mahavrila replied:
I do agree, but I have no clear idea of what to check except that ca._ca_cert exists. CustomCA setup is simple and straightforward. If you have an idea for more specific check I will be happy to implement it.

x00Pavel commented on SCAutolib/controller.py:
```diff
@@ -221,136 +253,132 @@ def setup_user(self, user_dict: dict, force: bool = False):
+        if force and card_dir.exists():
+            rmtree(card_dir)
+
+        if card_dict["card_type"] == "physical":
```
better to put this hardcoed string `"physiscal"` to the enum like
```python
class CardType(str, Enum):
   physical = "physiscal"
   virtual = "virtual"
```
and then use these values like `CardType.physical` because this approach would make your code more readable, explictely specified types, easy-to-change types

x00Pavel commented on SCAutolib/controller.py:
```diff
@@ -221,136 +253,132 @@ def setup_user(self, user_dict: dict, force: bool = False):
+        elif card_dict["card_type"] == "virtual":
+            hsm_conf = file.SoftHSM2Conf(card_dir.joinpath("sofhtsm2.conf"),
+                                         card_dir=card_dir)
             hsm_conf.create()
             hsm_conf.save()
 
-            new_user.card = card.VirtualCard(new_user,
-                                             softhsm2_conf=hsm_conf.path)
+            new_card = card.VirtualCard(card_dict, softhsm2_conf=hsm_conf.path,
+                                        card_dir=card_dir)
+
+            # card needs to know some details of its user, so we add user as
+            # card attribute
+            for card_user in self.users:
+                if card_user.username == card_dict["cardholder"]:
+                    new_card.user = card_user
+
+            if new_card.user.user_type == "local":
+                cnf_path = new_card.card_dir.joinpath(f"{new_card.cardholder}.cnf")
+                cnf = file.OpensslCnf(filepath=cnf_path, conf_type="user",
+                                      replace=[new_card.cardholder, new_card.CN])
+                cnf.create()
+                cnf.save()
+
+                new_card.cnf = cnf.path
+
+                if force:
+                    if new_card.cert and new_card.cert.exists():
+                        self.local_ca.revoke_cert(new_card.cert)
+            else:  # IPA user
+                if force:
+                    if new_card.cert and new_card.cert.exists():
+                        self.ipa_ca.revoke_cert(new_card.cert)
```
it would be nice to reduce the size of this single method and move this code to separate function just for better separation of concerns and potential testability

x00Pavel commented on SCAutolib/models/CA.py:
```diff
@@ -80,23 +103,33 @@ def load(json_file):
             cnt = json.load(f)
 
-        if "_ipa_server_ip" in cnt.keys():
+        if cnt["ca_type"] == "IPA":
```
same as with card type, enum for ca types would be better then hardcoded string

x00Pavel commented on SCAutolib/models/CA.py:
```diff
@@ -106,15 +139,19 @@ def __init__(self, root_dir: Path = None, cnf: OpensslCnf = None):
         self.root_dir: Path = Path("/etc/SCAutolib/ca") if root_dir is None \
-            else root_dir
+            else Path(root_dir)
         assert self.root_dir is not None
```
instead of assert, better to use explicet raise of an excpetion

x00Pavel commented on SCAutolib/models/card.py:
```diff
@@ -161,10 +183,18 @@ def __exit__(self, exp_type, exp_value, exp_traceback):
 
     def to_dict(self):
         return {
```
checkt what built-it method `__dict__` would return, maybe it would smth you need

mahavrila replied:
Thanks for the suggestion, at this moment, I want to keep the content under control by assigning values manually.

x00Pavel commented on SCAutolib/models/card.py:
```diff
@@ -275,11 +296,139 @@ def create(self):
+        `openssl` command based on template CNF file.
+        """
+        csr_path = self.card_dir.joinpath(f"csr-{self.cardholder}.csr")
+        if self.user.user_type == "local":
```
again, enum instead of hardcoded string with user type

x00Pavel commented on SCAutolib/models/user.py:
```diff
@@ -268,17 +122,14 @@ def __init__(self, ipa_server: IPAServerCA, *args, **kwargs):
         """
 
         super().__init__(*args, **kwargs)
+        self.user_type = "ipa"
```
use enum for user type

x00Pavel commented on SCAutolib/utils.py:
```diff
@@ -183,29 +206,49 @@ def ipa_factory():
+
+    if not path:            # create CA for physical card
+        ca = CustomCA(card_data)
+        ca.setup()
```
return something from the branch :)

mahavrila commented:
Hello @GeorgePantelakis , @Jakuje  , @x00Pavel. Updated version of code reflecting most of previous comments is ready. Note, that unit tests are failing and I did not updated them to pass with V3 as some of them are destructive (changes system config files etc.) and should probably be reviewed and rewritten.

Jakuje commented on SCAutolib/controller.py:
```diff
@@ -221,136 +275,140 @@ def setup_user(self, user_dict: dict, force: bool = False):
+        dump_to_json(card)
 
-    def cleanup(self):
+    def cleanup(self):  # FIXME
```
its a good practice to describe the fixme with at least one sentence so the next reader understand what is the issue and what needs to be fixed (even if not how).

Jakuje commented on SCAutolib/models/card.py:
```diff
@@ -21,22 +22,28 @@ class Card:
+            self.uri = urls[0]
+            logger.info(f"Card URI is set to {self.uri}")
+        elif len(urls) == 0:
+            logger.info("URI matching expected pattern not found.")
+            logger.warning("URI not set")
+            raise SCAutolibException("URI matching expected pattern not found.")
```
Can we simplify this logging to one log and one exception or do we need all?

GeorgePantelakis commented on SCAutolib/models/card.py:
```diff
@@ -92,34 +97,61 @@ class VirtualCard(Card):
-        self.dump_file = LIB_DUMP_CARDS.joinpath(
-            f"card-{self._user.username}.json")
+        self._nssdb = self.card_dir.joinpath("db")
+        self.dump_file = LIB_DUMP_CARDS.joinpath(f"{self.name}.json")
```
Double entry of dump_file. Same as line 143.

GeorgePantelakis reviewed (commented):
Not much changes. Mostly some nitpicking changes.

mahavrila commented on SCAutolib/controller.py:
```diff
@@ -221,136 +275,143 @@ def setup_user(self, user_dict: dict, force: bool = False):
+                new_card.cnf = self.prepare_user_cnf(new_card)
+            if force:
+                self.revoke_certs(new_card)
+            new_card.create()
```
I assume that in such case it will use existing certificates - i.e. it will not create new certs later on. This could be useful if for instance one wants to use specific certs. These certs will not be erased if force=False.

mahavrila commented on SCAutolib/models/CA.py:
```diff
@@ -106,15 +141,20 @@ def __init__(self, root_dir: Path = None, cnf: OpensslCnf = None):
+        :param cnf: object representing openssl cnf file
+        :type cnf: OpensslCnf
         """
+        self.name = LocalCA.ca_name
```
I wanted to declare ca_name that is the same for all instances of the class, however, later in code we use "__dict__.items" and If I remember correctly, this did not work on class varialbles so I have to add variable also to __init__. There we have self.name.

mahavrila commented on SCAutolib/models/CA.py:
```diff
@@ -343,6 +409,8 @@ def __init__(self, ip_addr: str, server_hostname: str, domain: str,
         :type realm: str
         """
+        self.ca_type = IPAServerCA.ca_type
+        self.name = IPAServerCA.ca_name
```
See previous reply.

mahavrila commented on SCAutolib/utils.py:
```diff
@@ -144,20 +133,44 @@ def user_factory(username, **kwargs):
     user_file = LIB_DUMP_USERS.joinpath(f"{username}.json")
-    logger.debug(f"Loading user {username} from {user_file}")
-    result = None
+    logger.debug(f"User {username} will be loaded from {user_file}")
```
I kept it here at the beginning for debugging purposes as controller calls various functions and while debugging early errors I could not see where exactly it happens. Loading of user happen in another method. I find this message useful in some weird way and tend to keep it. "Loading user" should probably be printed from user.load  method.

mahavrila replied:
Changed it back as suggested.

mahavrila commented on SCAutolib/utils.py:
```diff
@@ -144,20 +133,44 @@ def user_factory(username, **kwargs):
+    :rtype: Card
+    """
+    card_file = LIB_DUMP_CARDS.joinpath(f"{card_name}.json")
+    logger.debug(f"Loading card {card_name} from {card_file}")
```
Fixed previous one to keep them same.

GeorgePantelakis commented:
The code has been already merged. Closing PR.
